### PR TITLE
Add Metal 4 M5 prefill optimizations

### PR DESCRIPTION
**Pull request by ivanfioravanti**
## Summary
- enable M5-class Metal 4 MPP prefill paths for Q8_0 dense matmuls, attention-output low projection, and staged routed MoE projections
- promote correctness-gated routed MoE boundaries: down from layer 2, gate/up from layer 13
- add a fused six-expert routed MoE sum kernel for the common top-k=6 prefill shape
- keep experimental probes and escape hatches for ablation (`DS4_METAL_MPP_DISABLE=1`, `DS4_METAL_MOE_SUM6_DISABLE=1`, staged routed MoE envs)

## Benchmarks
Prompt source: `README.md`; command shape: `./ds4 --prompt-file <prompt> -n 1 --nothink --ctx 32768`; 3 repeats per target.

| target tokens | standard non-M5 avg tok/s | M5 no sum6 avg tok/s | M5 default avg tok/s | M5 speedup |
|---:|---:|---:|---:|---:|
| 512 | 236.97 | 390.29 | 389.47 | 1.64x |
| 2048 | 324.90 | 480.07 | 478.36 | 1.47x |
| 4096 | 289.97 | 454.41 | 455.52 | 1.57x |
| 8192 | 287.60 | 440.43 | 442.69 | 1.54x |
| 16384 | 284.26 | 427.62 | 427.78 | 1.50x |

8192-token routed MoE stage profile, current M5 default:

| stage | total ms | mean per layer ms | share |
|---|---:|---:|---:|
| up | 2443.497 | 11.365 | 35.1% |
| gate | 2422.942 | 11.269 | 34.8% |
| down | 1727.114 | 8.033 | 24.8% |
| activation_weight | 142.693 | 0.664 | 2.1% |
| sum | 137.482 | 0.639 | 2.0% |
| map | 68.601 | 0.319 | 1.0% |
| gate_up | 12.389 | 0.058 | 0.2% |

Disabling routed MoE MPP on the same profile drops prefill from 442.33 tok/s to 359.17 tok/s and raises gate/up/down to about 18.4-18.7 ms per layer.

## Validation
- `make ds4 ds4_test`
- `./ds4_test --metal-kernels`
- `./ds4_test --long-context`
- `./ds4_test --logprob-vectors`

## Notes
- Earlier `up` boundaries with `gate=13` were tested but failed long-context; gate/up stay paired at layer 13.
- A paired MPP gate+up matmul prototype compiled and ran but was slower overall, so it was not included.
- `DS4_METAL_MOE_MID_F32=1` looked slightly faster in a noisy local check, but the result is too small to promote without a broader clean sweep.

### CODE REVIEW

**Comment by antirez** [comment]
Very significant speedup. I'm in the middle of a large refactoring so can't merge right now and there is the problem that I miss the hardware to later maintain it, but I'll see what I can do :) Thanks.

**Comment by ivanfioravanti** [comment]
100% with you, we need to sort this out and let you get an M5 😎

**Comment by ivanfioravanti** [comment]
This should solve #14

**Comment by antirez** [comment]
@ivanfioravanti potential idea to get this merged: we take a m5-metal4 branch active, and you try to keep it rebased, if you like the idea. And we document it.

**Comment by ottaviofogliata** [comment]
@ivanfioravanti, just jumping in :) In a couple of days I’ll switch to a M5 Max 128GB. If it could be useful, I’d be more than happy to help you maintain the branch.

**Comment by ivanfioravanti** [comment]
Oh yes @ottaviofogliata join the club! I'm trying to squeeze even more juice with the various optimizations suggested in the [Metal Performance Primitives (MPP) Programming Guide](https://developer.apple.com/download/files/Metal-Performance-Primitives-Programming-Guide.pdf) without luck so far.

**Comment by ivanfioravanti** [comment]
I squeezed a little more juice. Tomorrow I'll test with pi mono for some coding sessions and I will also get server side stats instead of client side.

comparison_speed_chart

**Comment by ivanfioravanti** [comment]
logits are slightly different than the ones created with --quality. Converting to draft while I investigate, greedy is perfect, but distribution behind is different.

**Comment by ucjonathan** [comment]
If someone at Apple knew what was good for their hardware sales, they would have an M5 studio 256GB and and M5 Macbook Pro 128GB on @antirez desk Monday afternoon.  Unfortunately I don't know anyone at Apple, but hopefully there are some developers at Apple watching this project that will wake up to this opportunity.

**Comment by ivanfioravanti** [comment]
@ucjonathan I was going to propose the same thing!

**Comment by lobanov** [comment]
~~For some strange reason I see no speedup when testing this branch on M5 Max 128GB. I ran two different commands:~~

Ignore, I've got confused with branches. I'm getting `ds4: prefill: 513.56 t/s, generation: 30.45 t/s` when running `./ds4 --prompt-file README.md --ctx 32768`. Amazing work!

**Comment by audreyt** [comment]
**Forward-compat heads-up while you're rebasing:** there's a latent compressor-APE interaction between this PR and Q8_0 ape ingestion (e.g. for stock-recipe GGUFs like cyberneurova's). On M5 Max:

1. `kernel_dsv4_compressor_store_one` (in `metal/dsv4_kv.metal`, upstream) only handles F16/F32 ape — Q8_0 falls through to the F32 branch and reads garbage.
2. A naive Q8_0 GPU dequant kernel for the prefill APE byte-strided path produces silently wrong output on M5 Max for compressor APE shapes (4 rows × 1024 cols). CPU-side dequant into a *per-call private* `MTLBuffer` is what works — a shared scratch loses writes when the same buffer is reused inside one command buffer (Metal kernels run in encode order; CPU writes don't participate in that ordering when the scratch is reused).

The bug only manifests when both your MPP work and Q8_0 ape branches are present, so nothing to fix in this PR in isolation. Just flagging in case #60 (the stock-recipe loader PR) lands — the fix at [`audreyt/ds4` `m5-support-q8_0-token-embd`](https://github.com/audreyt/ds4/tree/m5-support-q8_0-token-embd) (commit [`79b08bb`](https://github.com/audreyt/ds4/commit/79b08bb)) would need to land alongside it. Happy to coordinate.

===
UPDATE: I made a [abliterated+aligned imatrix v2 quabt](https://huggingface.co/audreyt/CyberNeurova-DeepSeek-V4-Flash-abliterated-GGUF/blob/main/cyberneurova-DeepSeek-V4-Flash-abliterated-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix-aligned.gguf) so #60 is no longer in play.

**Comment by ivanfioravanti** [comment]
Thanks for flagging this. Currently I facing logprob drift in this PR compared to standard metal kernel and CPU, not big but I'm trying to lower it as much as possible. I will then rebase to main and start testing with some coding harness. 
This model rocks and this engine too!

**Comment by ivanfioravanti** [comment]
logprob drift fixed 🎉 Now I'm more confident from mathematical perspective!
Today I'll test performance, run some evals and try to use this branch with coding agents and hermes. 💪

**Comment by ivanfioravanti** [comment]
@audreyt I will port the MPP additions here to the new naming on main branch. I will test this PR like crazy after that and then rebase.

**Comment by ivanfioravanti** [comment]
We are getting there. I've used ds4-bench to make things easier to compare.

Don't like the MPP name (Metal Performance Primitives), let's use something more related to the new Tensor APIlike Metal Tensor (mt).

There is a difference in logprob between standard metal kernel and this one, so I will now run evals on both to measure impacts. If this is ok, I will ask super @antirez how to proceed.

ds4-bench-standard-vs-mpp

**Comment by ivanfioravanti** [comment]
I did a full AIME2025 evaluation round of standard metal kernel vs tensor metal with max context 16K and I've got:
- 81.5% Standard
- 73.3% Tensor

I'll keep this PR align to main and keep reducing logprob diff while ensuring a performance boost at the same time. 💪

**Comment by ivanfioravanti** [comment]
## Status update

### Quality perspective
Now Tensor Metal Kernel and Standard Metal Kernel are very similar from mathematical perspective and generation equivalent, here the comparison of each vs --quality. 

Tensor Metal:
  short_italian_fact      same_top1=yes  top20=20/20  rms=0.370521  top20_abs=0.659939
  short_code_completion   same_top1=yes  top20=19/20  rms=0.463208  top20_abs=2.24006
  short_reasoning_plain   same_top1=yes  top20=18/20  rms=0.618172  top20_abs=1.98223
  long_memory_archive     same_top1=yes  top20=19/20  rms=0.578821  top20_abs=1.10567
  long_code_audit         same_top1=yes  top20=18/20  rms=0.428341  top20_abs=1.04939

Default Metal (Main branch)
  short_italian_fact      same_top1=yes  top20=20/20  rms=0.370521  top20_abs=0.659939
  short_code_completion   same_top1=yes  top20=19/20  rms=0.463208  top20_abs=2.24006
  short_reasoning_plain   same_top1=yes  top20=18/20  rms=0.618172  top20_abs=1.98223
  long_memory_archive     same_top1=no   top20=19/20  rms=0.380069  top20_abs=0.692255
  long_code_audit         same_top1=yes  top20=16/20  rms=0.530931  top20_abs=1.38902

Here the test against the other M5 implementation in [Swival ds4-m5](https://github.com/Swival/ds4-m5)
  short_italian_fact      same_top1=yes  top20=19/20  rms=0.355529  top20_abs=1.20708
  short_code_completion   same_top1=yes  top20=16/20  rms=0.689058  top20_abs=1.58729
  short_reasoning_plain   same_top1=yes  top20=20/20  rms=0.556018  top20_abs=1.50008
  long_memory_archive     same_top1=yes  top20=16/20  rms=0.510137  top20_abs=1.65245
  long_code_audit         same_top1=yes  top20=17/20  rms=0.612309  top20_abs=1.23852

From eval perspective the new slower Tensor Metal Kernel has improved AIME2025 reaching 90% in Q2 (unbelievable!)
image

### Performance
Here situation is clearly different from previous version, there is still a benefit but more in the range of ~15%.
There is room from improvement and I'll keep pushing this, but with strict bar on logprob drift.

ds4_bench_standard_quality_tensor_128

@antirez I'll leave this as draft and I'll keep rebasing on main, I'll remove the drift prone Q8_0 Tensor route and all related tests, docs, etc.

When you'll get your M5 Max gifted by super @audreyt I'm sure you'll be able to do even more.

**Comment by variablefate** [comment]
Quick question on 670411d. `ds4_hc_sigmoid` and `ds4_hc_twice_sigmoid` get called from `kernel_dsv4_hc_split_sinkhorn`, but `kernel_dsv4_hc_split_weighted_sum` (lines 304/314) and `kernel_dsv4_hc_split_weighted_sum_norm4` (lines 429/439) still use inline `1/(1+exp(-z))`. The norm4 variant is on the decode fast path per the comment at `ds4_metal.m:15926-15928`. Was the scoping to sinkhorn for tanh-vs-exp throughput reasons on the fused kernels, did your bisection localize the drift there specifically, or are you still planning on running those four through the helpers?

**Comment by ivanfioravanti** [comment]
Good catch @variablefate This was scoped intentionally to the standalone Sinkhorn split path used by the prefill/logit-drift harness. I did not move the fused decode paths over because they are on the generation fast path, and at that point I was trying to isolate the prefill drift without changing decode math or risking a tanh-vs-exp throughput hit there.

I’m now working on re-adding selected Tensor/Metal paths and pushing performance back up, while keeping a strict bar for mathematical adherence to the original Metal kernel. Once the prefill side is stable enough, I can also A/B the fused decode sigmoid helpers separately and only keep them if the quality/perf tradeoff is clearly worth it.

**Comment by ivanfioravanti** [comment]
Ok I'm finally happy with new results and ready to update the PR and remove the draft status.

## Performance
From performance perspective we are in good shape as you can see in the chart.

image

## Quality
From quality perspective I've got some eval results on AIME 2025 that is 86.7% using q2-imatrix and 65K max-tokens on standard metal and tensor metal kernel.

I have added reusable local validation tooling under `speed-bench/` for Metal Tensor work:

  - `run_quality_drift_gate.py` runs the fixed five-scenario logprob drift comparison across `--quality`,
  standard Metal (`-mt off`), and Tensor Metal (`-mt auto`), writing both `summary.md` and `summary.json`
  for PR-ready reporting.
  - `run_metal_tensor_bench.sh` regenerates the Standard vs Quality vs Tensor performance chart with
  timestamped CSV/PNG artifacts under ignored `speed-bench/local-runs/`.
  - `run_prefill_candidate_gate.py` provides a speed-first promotion gate for default-off Tensor prefill
  candidates, then runs drift checks only when the speed screen passes.
  - `index_local_runs.py` indexes saved local benchmark, drift, comparator, and profile artifacts so
  results can be compared across future runs.

These tools keep benchmark artifacts local by default and make the PR evidence reproducible without committing raw run outputs.

### Logprob drift under control
tensor_vs_standard
case same_top1 top5 top20 rank rms max_abs top20_abs greedy
short_italian_fact yes 5/5 20/20 0 0 0 0 same
short_code_completion yes 5/5 20/20 0 0 0 0 same
short_reasoning_plain yes 5/5 20/20 0 0 0 0 same
long_memory_archive yes 5/5 19/20 1 0.166312 0.868076 0.55422 same
long_code_audit yes 5/5 20/20 5 0.239946 1.2248 0.330135 same
summary top1_mismatches=0 greedy_mismatches=0 min_top20=19/20 worst_rms=0.239946 worst_top20_max_abs=0.55422

**Comment by ivanfioravanti** [comment]
Update after eval drift check:

The previous routed-MoE Tensor default enabled the up projection from layer 36. That passed the q1-q12 gate used in #169 but q11 exposed a deterministic drift:
- old default: q11 generated 1725 tokens and the current eval extractor reported `I/A`
- with `DS4_METAL_MPP_MOE_UP_START_LAYER=37`: q11 matches main exactly:
    `prompt=137 gen=851 total=988 given=A correct=A`

This commit moves only the default routed-MoE up Tensor start layer from 36 to 37. Gate/down stay at layer 35.

Prefill speed is lower, but mathematically identical to standard kernel. 
I will apply this same deterministic test to llamacpp and mlx_lm to see if they bring same results.

**Comment by fitchmultz** [comment]
FYI I opened #185 as a small eval-harness/audit PR. It adds `ds4-eval --regrade-trace` and makes the extractor prefer the final strict `Answer:` line, without changing inference.

Why it may matter for this drift discussion: regrading my saved pre-equality #15 trace shows q11's final visible line was `Answer: A`, but the old extractor reported `I`. So that q11 letter mismatch was an evaluator artifact. The token-count/logit drift in that run is still a separate real signal.

**Comment by ivanfioravanti** [comment]
That's great @fitchmultz 🙌🏻
I will rerun evals using #185 results are encouraging for the old default but faster (on prefill) version of this PR. I will rollback it and rebase to main if evals are good. So far they look pretty strong. 
I've tried to add your your boost on decoding, but no luck.

**Comment by ivanfioravanti** [comment]
Rebased to main. Feel free to pick or change anything now that you have M5 to try this out @antirez 💪

**Comment by antirez** [comment]
Perfect. Thanks.

**Comment by antirez** [comment]
M5 Max / Metal4 support merged. The prefill speed is very interesting, I added more optimziations on top of those and now with small contexts it is extremely fast. Thanks for all the work.
